### PR TITLE
Get rid of is_min_const_fn

### PR DESCRIPTION
**Pull request by RalfJung**
This removes the last trace of the min_const_fn mechanism by making the unsafety checker agnostic about whether something is a min or "non-min" const fn. It seems this distinction was used to disallow some features inside `const fn`, but that is the responsibility of the const checker, not of the unsafety checker. No test seems to even notice this change in the unsafety checker so I guess we are good...

r? @oli-obk 
Cc https://github.com/rust-lang/rust/issues/84510

### CODE REVIEW

**Comment by RalfJung** [comment]
Some checks used `::General` before (and thus would have lead to an error in `min_const_fn`):
* inline assembly: https://github.com/rust-lang/rust/blob/13a2615883aa28433383a723a764ca9acb43fd48/compiler/rustc_mir/src/transform/check_unsafety.rs#L98
* other inline assembly: https://github.com/rust-lang/rust/blob/13a2615883aa28433383a723a764ca9acb43fd48/compiler/rustc_mir/src/transform/check_unsafety.rs#L121
* Ptr-to-int cast: https://github.com/rust-lang/rust/blob/13a2615883aa28433383a723a764ca9acb43fd48/compiler/rustc_mir/src/transform/check_unsafety.rs#L160
* Static accesses
https://github.com/rust-lang/rust/blob/13a2615883aa28433383a723a764ca9acb43fd48/compiler/rustc_mir/src/transform/check_unsafety.rs#L196
https://github.com/rust-lang/rust/blob/13a2615883aa28433383a723a764ca9acb43fd48/compiler/rustc_mir/src/transform/check_unsafety.rs#L202

Those should all be rejected by our regular const checks, right?

**Comment by oli-obk** [comment]
> Those should all be rejected by our regular const checks, right?

Yea, we already have feature gates for them, that is independent

**Comment by oli-obk** [comment]
Can you double check that we still make some things unsafe in const fn that are safe in regular fns? (Like pointer comparisons)

**Comment by RalfJung** [comment]
> (Like pointer comparisons)

I don't think pointer comaprisons were ever unsafe? They should just be forbidden.
Ptr-to-int casts are unsafe. I extended the corresponding test in the last commit.

**Comment by RalfJung** [comment]
Raw ptr comparison is checked here

https://github.com/rust-lang/rust/blob/46d09f7a4b2fe6675f06394a187cef46728412d3/compiler/rustc_mir/src/transform/check_consts/validation.rs#L704

There is no feature gate for `RawPtrComparison`, it is just always forbidden... is it deliberate that `ops` are used both for "must never be allowed" and "is currently unstable"?

**Comment by oli-obk** [comment]
> is it deliberate that `ops` are used both for "must never be allowed" and "is currently unstable"?

Yes, otherwise unleash can't know what checks it can skip

**Comment by oli-obk** [comment]
r=me with ci passing

**Comment by RalfJung** [comment]
I am not sure if unleashing should allow raw ptr comparisons...

**Comment by RalfJung** [comment]
> r=me with ci passing

@bors r=oli-obk

File: compiler/rustc_middle/src/mir/query.rs [hunk 1]
```diff
@@ -19,10 +19,8 @@ use super::{Field, SourceInfo};
 
 #[derive(Copy, Clone, PartialEq, TyEncodable, TyDecodable, HashStable, Debug)]
 pub enum UnsafetyViolationKind {
-    /// Only permitted in regular `fn`s, prohibited in `const fn`s.
+    /// Unsafe operation outside `unsafe`
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/query.rs:22]:
```suggestion
    /// Unsafe operation outside `unsafe`.
```

**Comment by oli-obk** [comment]
> I am not sure if unleashing should allow raw ptr comparisons... but yeah makes sense.

We should skip the static check on raw pointer comparisons in unleash mode in order to demonstrate the runtime behaviour if there are actual relocations involved

**Pull request merged**